File: fs/src/test/scala/farjs/fs/MockFSService.scala
package farjs.fs

import scala.scalajs.js

//noinspection NotImplementedCode
class MockFSService(
  openItemMock: (String, String) => js.Promise[Unit] = (_, _) => ???,
  readDiskMock: String => js.Promise[js.UndefOr[FSDisk]] = _ => ???,
  readDisksMock: () => js.Promise[js.Array[FSDisk]] = () => ???
) extends FSService {

  override def openItem(parent: String, item: String): js.Promise[Unit] =
    openItemMock(parent, item)
    
  override def readDisk(path: String): js.Promise[js.UndefOr[FSDisk]] =
    readDiskMock(path)
    
  override def readDisks(): js.Promise[js.Array[FSDisk]] =
    readDisksMock()
}

File: fs/MockFSService.mjs
/**
 * @import { FSService } from "./FSService.mjs"
 */

/**
 * @typedef {{
 *  openItem?: FSService['openItem'];
 *  readDisk?: FSService['readDisk'];
 *  readDisks?: FSService['readDisks'];
 * }} FSServiceMocks
 */

/**
 * @implements {FSService}
 */
class MockFSService {
  /**
   * @param {FSServiceMocks} mocks
   */
  constructor({ openItem, readDisk, readDisks } = {}) {
    this.openItem = openItem ?? this.openItem;
    this.readDisk = readDisk ?? this.readDisk;
    this.readDisks = readDisks ?? this.readDisks;
  }

  /** @type {FSService['openItem']} */
  openItem() {
    return Promise.reject(new Error("Not implemented!"));
  }

  /** @type {FSService['readDisk']} */
  readDisk() {
    return Promise.reject(new Error("Not implemented!"));
  }

  /** @type {FSService['readDisks']} */
  readDisks() {
    return Promise.reject(new Error("Not implemented!"));
  }
}

export default MockFSService;
